### DISCUSSION TITLE
Clip from an local pdf embedded in a tiddler into  stream tiddlers with same parent tiddler

### DISCUSSION THREAD

**Discussion by arunnbabu81** [buggyj/tiddlyclip-plugin#107] [Ideas]
https://demos.tiddlyhost.com/#Open%20pdf%20tiddler%20at%20page%20X%2FDemo

This is a pdf macro created by telumire in talk tiddlywiki forum which can be used to open a specific page in an pdf embedded in a tiddler. 

Can I clip text or images from such local pdfs embedded in a tiddler using tiddlyclip. I know that the tid feature is currently not fully functional as told here. 

https://github.com/buggyj/tiddlyclip-plugin/issues/81#issuecomment-1059719846

Will this functionality to clip from a tiddler in the same wiki be back in tiddlyclip.

If so I can make use of the pdf macro shown above and make notes with a link to open the corresponding page in the emebedded pdf.

**Comment by buggyj** [comment]
The tid feature copies tiddlers from one tiddlywiki to another, it does not have anything to do with what you want to do.  I do not know if you can directly clip text from an embedded pdf.

**Comment by arunnbabu81** [comment]
If so, will snip or snap option work in the way I was describing.

**Reply by buggyj** [reply]
probably.

**Reply by arunnbabu81** [reply]
> probably.

Snip and snap options are working.
Tiddler with the embedded pdf usually have a suffix `.pdf` in its title
 I want the title of the clipped tiddler be the same as the pdf tiddler except for the suffix `.pdf`

How to acheive that?

**Reply by buggyj** [reply]
It is possible to 'dock' to individual tiddlers - I will upload a button for you

**Reply by buggyj** [reply]
Are you putting the clips into a normal or a steam tiddler?

**Reply by arunnbabu81** [reply]
I use both normal and stream tiddlers depending upon the type of content.

**Reply by buggyj** [reply]
First you must be docked to the tiddlywiki you are using. 
Then in the view buttons on the tiddler top  you will find a tiddlyclip button to dock to that tiddler (eg xxxx.pdf), a new tiddler without the .pdf will be created for the clips (eg xxxx)

[localtcplugin.json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/10085833/localtcplugin.json.zip)

**Comment by arunnbabu81** [comment]
Its working. But I am seeing this error message whenever I click anywhwere within the wiki 

<img width="1082" alt="image" src="https://user-images.githubusercontent.com/67494083/203852355-94e5abea-8f5e-48b3-9d70-ade3a4a6c917.png">
It must be due to conlict with some of my plug ins or tweaks. Shall I upload a demo wiki ? I could not reproduce the error after doing a fresh installation-will investigate if if recurrs.

Also will it not work in chrome. I got the error message in firefox, but I wanted to whether the error happens in chrome also- but in chrome the clipping was not working. Both chrome and firefox has same version of the add on

Also how to add more configurations to the localtcplug in. I want to do snap, some wiktext formatting and add support for streams, pdf macro.

**Reply by buggyj** [reply]
it seems that you cannot clip text from an embedded pdf in chrome

**Reply by arunnbabu81** [reply]
> Also how to add more configurations to the localtcplug in. I want to do snap, some wiktext formatting and add support for streams, pdf macro.

Is this possible

**Reply by buggyj** [reply]
the config for the docked tiddler is `PdfClipConfig`. Note that `@section` is the title of the docked tiddler.

**Comment by buggyj** [comment]
the config for the docked tiddler is ''PdfClipConfig''. Note that ''@section'' is the title of the docked tiddler.

**Reply by arunnbabu81** [reply]
> the config for the docked tiddler is ''PdfClipConfig''. Note that ''@section'' is the title of the docked tiddler.

I got most of the configurations right. But I unable to create the correct configuration for the child node of streams tiddlers. Can you help. Here is the demo - https://localtiddlyclip-stream.tiddlyhost.com/

I dont know what to do for the `parent` variable in the field initials section and in the `streaminc` tiddler

**Reply by buggyj** [reply]
in tid 'LocalTitle' make this change

```
#rule=/^(.*?).pdf$/g
#term=$1
$title=((*@section/#rule/#term*))
%parent=((*$title*))
```

Then change 
`
{"$parent":"[[LocalTitle]]"}`

to

`{"$parent":"((*%parent*))"}`


and 

`$stream-list=((*@fixme($stream-list,$stream-list)*)) [[LocalTitlestreaminc]]`

to 

`$stream-list=((*@fixme($stream-list,$stream-list)*)) [[((*%parent*))/sub((*%new*))]]`

**Reply by buggyj** [reply]
There is as small fix for the plugin:
[localtcplugin.json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/10095814/localtcplugin.json.zip)

**Reply by arunnbabu81** [reply]
Finally got everything to work. Thank you @buggyj

**Comment by arunnbabu81** [comment]
Can the config tiddler of the localtcplug-in (pdfclipconfig) be updated to have better naming in the context menu. Currently the name `pdfclipconfig` is seen in all submenus of the context menu corresponding to te different categories I created in the pdfclipconfig tiddler. While this plug in was created for pdf clipping, it can be used to adding snips or snaps or pin to already existing tiddler in other ways also.

**Reply by buggyj** [reply]
`pdfclipconfig` does not appear in my context menu - can you post a screen shot?

**Reply by buggyj** [reply]
or update you example online wiki

**Reply by arunnbabu81** [reply]
<img width="1605" alt="image" src="https://user-images.githubusercontent.com/67494083/206921524-28b42115-547e-4951-bc94-694f06c45c1d.png">


<img width="1611" alt="image" src="https://user-images.githubusercontent.com/67494083/206921553-b85f565a-d8d6-4c81-81a5-c81e93e69862.png">

**Reply by arunnbabu81** [reply]
I hadn't tested those rules after I created those categories in the `pdfclipconfig` tiddler. Now after I tested I understood that those rules were not working after I made that change to the `pdfclipconfig` tiddler. So I reverted back to the old state. Now the `pdfclipconfig`  is not seen in the context menu. Sorry for the confusion.

<img width="1824" alt="image" src="https://user-images.githubusercontent.com/67494083/206922851-d0d5990b-a5a0-4c13-8d7f-af1efc5a7043.png">

The clips was not getting clipped to the docked tiddler after I made the changes to the `pdfclipconfig` tiddler. Whats is the reason for that. It would be nice to create subcategories within the `pdfclipconfig` tiddler.

**Reply by buggyj** [reply]
it would be possible to have more than one `section`.

it should also be possible to have different configs for different types of tiddlers.

**Reply by arunnbabu81** [reply]
<img width="1820" alt="image" src="https://user-images.githubusercontent.com/67494083/206925495-36d47e29-50be-4be7-90eb-30fb233e8987.png">

This one is working, but we cannot make out the subcategory from the context menu.

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/67494083/206925528-d293ce4c-cad7-43e8-9179-4534b415fce2.png">
If I add a text to the ((*@section*)) like ((*@section*))Default or ((*@section*))pdf , that text also gets added to the clipped tiddler title.

**Reply by buggyj** [reply]
The 'LocalTitle' tid can be modified to remove the extra text.

```
#rule0=/^(.*) .*?$/g
#rule=/^(.*?).pdf$/g
#term=$1
#section=((*@section/#rule0/#term*))
$title=((*#section/#rule/#term*))
```

But you need a space between `((*@section*))` and 'default' in the PdfClipConfig
`
((*@section*)) default`

**Reply by arunnbabu81** [reply]
> The 'LocalTitle' tid can be modified to remove the extra text.
> 
> ```
> #rule0=/^(.*) .*?$/g
> #rule=/^(.*?).pdf$/g
> #term=$1
> #section=((*@section/#rule0/#term*))
> $title=((*#section/#rule/#term*))
> ```
> 
> But you need a space between `((*@section*))` and 'default' in the PdfClipConfig
> `
> ((*@section*)) default`

Thank you @buggyj I will test it when I am back on my desktop.

**Reply by buggyj** [reply]
A better solution is to store the name of the docked tiddler in temporary tiddler.

**Reply by arunnbabu81** [reply]
> A better solution is to store the name of the docked tiddler in temporary tiddler.
> 
> 

How to do that

**Reply by buggyj** [reply]
I am updating the plugin with this new method and will explain it when I upload it

**Reply by buggyj** [reply]
[localtcplugin(1).json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/10229609/localtcplugin.1.json.zip)

So this now stores the name of the docked tiddler, when the tiddlyclip button is clipped. 
When the tclip button is pressed a tiddlyclip-do action is called to action `$:/bj/tidDockRule` which stored the name of the docked tiddler in the temp tid called `$:/temp/tiddock`. Later, when clipping, the name of the docked tiddler is retrieved from this temp tid, eg in the tiddler `LocalTitle` can be seen the line:

`#title:=$:/temp/tiddock` 

note that the pair of characters `:=` cause the contents  of  tiddler to be loaded into `#title`.


There are now separate config tiddlers for pdf and normal tiddlers -



type | config
-- | --
application/pdf | PdfClipConfig
text/vnd.tiddlywiki | TidClipConfig

**Reply by arunnbabu81** [reply]
> [localtcplugin(1).json.zip](https://github.com/buggyj/tiddlyclip-plugin/files/10229609/localtcplugin.1.json.zip)
> 
> So this now stores the name of the docked tiddler, when the tiddlyclip button is clipped. When the tclip button is pressed a tiddlyclip-do action is called to action `$:/bj/tidDockRule` which stored the name of the docked tiddler in the temp tid called `$:/temp/tiddock`. Later, when clipping, the name of the docked tiddler is retrieved from this temp tid, eg in the tiddler `LocalTitle` can be seen the line:
> 
> `#title:=$:/temp/tiddock`
> 
> note that the pair of characters `:=` cause the contents of tiddler to be loaded into `#title`.
> 
> There are now separate config tiddlers for pdf and normal tiddlers -
> 
> type	config
> application/pdf	PdfClipConfig
> text/vnd.tiddlywiki	TidClipConfig

Thank you @buggyj Now its perfect.
